src/chainstate/burn/sortition.rs

```diff
@@ -173,147 +163,220 @@ impl BlockSnapshot {
-            //      -- we either did a sortition last block, OR
-            //      -- no blocks were committed (so no sortition) and the burn quota was 0
-            // -- we DECREMENT otherwise
-            if last_sortition || (last_sortition_burn_total == 0 && last_burn_quota == 0) {
-                debug!("SORTITION({}): burned enough for sortition last block, so QUOTA INCREMENT", block_height);
-                next_burn_quota = BlockSnapshot::burn_quota_inc_checked(last_burn_quota, burnchain.burn_quota.inc);
-            }
-            else {
-                debug!("SORTITION({}): did no burn enough for sortition last block, so QUOTA DECREASE", block_height);
-                next_burn_quota = BlockSnapshot::burn_quota_dec_checked(last_burn_quota, burnchain.burn_quota.dec_num, burnchain.burn_quota.dec_den);
+        // total burn.  If this ever overflows, then just stall the chain and deny all future
+        // sortitions (at least the chain will remain available to serve queries, but it won't be
+        // able to make progress).
+        let next_burn_total = match last_burn_total.checked_add(block_burn_total) {
+            Some(new_total) => {
+                new_total
+            },
+            None => {
+                // overflow.  Deny future sortitions
+                warn!("Cumulative sortition burn has overflown.  Subsequent sortitions will be denied.");
+                return BlockSnapshot::make_snapshot_no_sortition(tx, parent_snapshot, block_header, first_block_height, last_burn_total, &next_sortition_hash, &txids);
             }
+        };
 
-            // we will have burned 0 towards the next sortition.
-            next_sortition_burn_total = 0;
-        }
-        else {
-            // cannot do a sortition in this block, but don't change the burn quota -- it will only
-            // be decremented once we meet it!  Instead, keep a running total of how much we burned
-            // since the last sortition, and use that in the next epoch to see if we can adjust the
-            // burn quota.
-            next_sortition = false;
-            next_burn_quota = last_burn_quota;
-            next_sortition_burn_total = sortition_burn_total;
-
-            winning_block_txid = Txid::from_bytes(&[0u8; 32]).unwrap();
-            winning_block_burn_hash = BurnchainHeaderHash::from_bytes(&[0u8; 32]).unwrap();
-
-            info!("SORTITION({}): Burn quota not met ({} < {}), so NO QUOTA CHANGE", block_height, block_burn_total, last_burn_quota);
-        }
-
-        let next_burn_total = BlockSnapshot::burn_total_checked(last_burn_total, block_burn_total);
-        
+        // Try to pick a next block.
+        let (winning_txid, winning_block_hash, winning_block_vrf_seed) = match BlockSnapshot::select_winning_block(tx, block_header, &next_sortition_hash, burn_dist)? {
+            None => {
+                // should be unreachable, but would happen if there were no burns
```

Comment:
Since this function checks for the 0 burn case above, should this branch panic? (or call .expects on the select_winning_block result)

Reply:
Yes -- this branch is unreachable because the only way that this method returns `None` is if `burn_dist` has length 0 (and this case is handled earlier in this method).

Reply:
Fixed in 8f8c6e10